pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -359,33 +360,128 @@ class StateVectorCudaManaged
+    void applyOperation(const std::string &opName,
+                        const std::vector<std::size_t> &controlled_wires,
+                        const std::vector<bool> &controlled_values,
+                        const std::vector<std::size_t> &tgt_wires,
+                        bool adjoint = false,
+                        const std::vector<Precision> &params = {0.0},
+                        const std::vector<ComplexT> &gate_matrix = {}) {
+        PL_ABORT_IF_NOT(
+            areVecsDisjoint<std::size_t>(controlled_wires, tgt_wires),
+            "`controlled_wires` and `target wires` must be disjoint.");
         PL_ABORT_IF(controlled_wires.size() != controlled_values.size(),
-                    "`ctrls` and `ctrls_values` must have the same size.");
+                    "`controlled_wires` and `controlled_values` must have the "
+                    "same size.");
         auto ctrlsInt = NormalizeCastIndices<std::size_t, int>(
             controlled_wires, BaseType::getNumQubits());
         auto tgtsInt = NormalizeCastIndices<std::size_t, int>(
             tgt_wires, BaseType::getNumQubits());
         auto ctrls_valuesInt =
             Pennylane::Util::cast_vector<bool, int>(controlled_values);
 
-        if (opName == "GlobalPhase") {
-            const std::vector<std::string> names(tgt_wires.size(), "I");
+        if (opName == "MultiRZ") {
+            const std::vector<std::string> names(tgtsInt.size(), {"RZ"});
+            applyParametricPauliGeneralGate_(names, ctrlsInt, ctrls_valuesInt,
+                                             tgtsInt, params.front(), adjoint);
+        } else if (opName == "GlobalPhase") {
+            const std::vector<std::string> names(tgtsInt.size(), "I");
             applyParametricPauliGeneralGate_(names, ctrlsInt, ctrls_valuesInt,
                                              tgtsInt, 2 * params[0], adjoint);
+
+        } else if (native_gates_.find(opName) != native_gates_.end()) {
+            applyParametricPauliGeneralGate_({opName}, ctrlsInt,
+                                             ctrls_valuesInt, tgtsInt,
+                                             params.front(), adjoint);
+        } else if (opName == "Rot") {
+            auto rot_matrix =
+                adjoint
+                    ? cuGates::getRot<CFP_t>(params[2], params[1], params[0])
+                    : cuGates::getRot<CFP_t>(params[0], params[1], params[2]);
+            applyDeviceGeneralGate_(rot_matrix.data(), ctrlsInt, tgtsInt,
+                                    ctrls_valuesInt, adjoint);
+        } else if (par_gates_.find(opName) != par_gates_.end()) {
+            // TODO: offload to par_gates_ if available
+            auto &gateMap =
+                cuGates::DynamicGateDataAccess<PrecisionT>::getInstance();
+            auto &&matrix_cu = gateMap.getGateData(opName, params);
+
+            gate_cache_.add_gate(opName, params[0], matrix_cu);
```

Comment:
Could you explain to us why we use here `auto &` and `auto &&`, please? We can not guess the reason for this. @josephleekl

Reply:
Good question! `DynamicGateDataAccess` follows the singleton implementation, which prevents us to return a rvalue of the instance. rvalue reference would violate the singleton pattern's principle of having only one instance. auto &&matrix_cu can avoid copies of std::vector and is faster